internal/api/proxy.go

```diff
@@ -0,0 +1,139 @@
+package api
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+
+	"golang.org/x/net/proxy"
+)
+
+func applyProxy(transport *http.Transport, proxyEndpointURL *url.URL, proxyEndpointPath string) (applied bool) {
+	if proxyEndpointURL == nil && proxyEndpointPath == "" {
+		return false
+	}
+
+	handshakeTLS := func(conn net.Conn, addr string) (net.Conn, error) {
+		// Extract the hostname (without the port) for TLS SNI
+		host, _, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, err
+		}
+		// Be sure to clone TLS-specific config settings from the transport,
+		// like InsecureSkipVerify.
+		tlsConn := tls.Client(conn, &tls.Config{
+			ServerName:         host,
+			InsecureSkipVerify: transport.TLSClientConfig.InsecureSkipVerify,
+		})
+		if err := tlsConn.Handshake(); err != nil {
+			return nil, err
+		}
+		return tlsConn, nil
+	}
+
+	var dial func(ctx context.Context, network, addr string) (net.Conn, error)
+	var dialTLS func(ctx context.Context, network, addr string) (net.Conn, error)
+
+	if proxyEndpointPath != "" {
+		dial = func(_ context.Context, _, _ string) (net.Conn, error) {
+			return net.Dial("unix", proxyEndpointPath)
```

Comment:
Also `net.DialContext` here (and everywhere else we call `Dial`)

Reply:
Those dial functions are applied to `transport.DialContext` and `transport.DialTLSContext`. They're just named without "Context" - poor name choice perhaps.

Reply:
Right! But the `net.Dial` call inside the function doesn't use the passed-in context, so will not respect any timeouts

Reply:
Oh, I see. Interestingly enough, there is no `net.DialContext` function - it's a method on `Dialer`, so I have to define a `Dialer`, then call `DialContext` on it. FWIW, `net.Dial` does the same thing.

Reply:
Oh interesting! That's surprising to me. TIL